### DISCUSSION TITLE
Extending the new admin console with custom tabs?

### DISCUSSION THREAD

**Discussion by sventorben** [keycloak/keycloak#13598] [Archive]
Is there a way to add additional tabs in the new admin console?

With the old admin console I was able to add custom tabs to the identity provider page. I did this by writing a custom theme and simply add an additional tab by adjusting a single Freemarker template file (ftl). 

See [here](https://github.com/sventorben/keycloak-home-idp-discovery/blob/v19.0.0/src/main/resources/theme/home-idp-discovery/admin/resources/templates/kc-tabs-identity-provider.html#L15) for an example:
![Custom Tab](https://github.com/sventorben/keycloak-home-idp-discovery/raw/v19.0.0/docs/images/view-idp-email-domains.jpg)

Is there any way to do something similar with the new admin console, either by adjusting the theme or some kind of extension mechanisms like the [content.js](https://github.com/keycloak/keycloak-quickstarts/blob/19.0/extend-account-console/src/main/resources/theme/keycloak-man/account/resources/content.json) for the account console?

**Comment by ssilvert** [comment]
Right now to do this you will need to fork the new admin console and create your own version.

We are thinking about how to make customization as easy possible and would love to hear any suggestions.

Have you found the content.js mechanism to be useful?  We are thinking of getting rid of it.

**Reply by sventorben** [reply]
Hello @ssilvert,

thanks for your reply. 
Yes, I found the content.js mechanism useful in the account console. It is a neat mechanism to add additional page there and show additional information to users. I am not sure, if this is what I need in the admin cosnole in my case - I was just looking around for potential extension mechanisms.

Currently Keycloak supports a lot of configuration options/properties for providers of any kind. This is somewhat a 1:1 relation between the config and the provider. What I would need in my case would be a mechanism to extend other provider's configuration. 

In my concrete example, I have a custom authenticator that redirects users to an IdP based on their email domain. I could store the mapping of email domains to IdP within the config of my authenticator. However, I thinks it would be nicer to be able to configure domains as part of the IdP config. 

Does this make sense to you?

**Reply by hypery2k** [reply]
the nice thing about the content.js method is, that you just have add the parts you want to add. I think in general there multiple levels of customization:

* *light*: Just add some more fields, e.g. for enforced user attributes naming, could be probably done via server config option
* *override*:  Tailor pages or tabs for custom usage, e.g. hide some details for a dedicated realm
* *custom*: Full custom admin theme with custom branding / UI-UX

I think option 1 would be easy to setup in combination with a backend provider and option 3 is already there ...

For option 2 I'm unsure how to do that. Maybe adding a "remote config" option to add custom "JS URLs" and then inject a web component into configured pages. Direct frontend overwrites won't work as the resources are hashed. So the only option would be to inject "a config block" from the server to override/customize certain areas ...

**Reply by vmichelin-cpage** [reply]
Hello,
I'm also interested by this evolution.

In my case, i've developper many user storage providers with theirs own templates and partials in the old admin console, and for the moment, i think it's not possible to extend the console.

@hypery2k : I my case, i could need the option 2 (for adding tabs for example) : would be nice to have a code server side description of the model, (List<ProviderConfigProperty> + UIProperties (tabs, groups, titles, ....) to render this automatically.

**Reply by hypery2k** [reply]
yes, that would be really nice. 
@ssilvert what do you think of this approach? Having a server side endpoint to add simple customizations there and then let the React frontend render these simple forms?

If you like I could send a PR for this ...

**Reply by antikalk** [reply]
I would also be interested in this one 👍

**Comment by ssilvert** [comment]
Thanks to everyone for your thoughts and suggestions.  It's extremely helpful to hear about your use cases and what you would like to see in the UI going forward.

Just as an update, the account console is being rewritten using modern React with an eye toward a customization strategy that will be reused in the admin console.  @edewit is heading up this effort.  Like admin console is today, it will still be somewhat of a "fork and modify" approach.  But there will be an emphasis on clean integration points and good developer documentation.  

The goal is that you will at least be able rely on the base "UI API" for a long time.  That way, your code will not be broken when we have new releases of the consoles.

Also, we are already using the kind of approach that @vmichelin-cpage suggested.  This is for the User Profile feature.  You define the user profile and the forms and validators are auto-generated in admin console and (soon) account console.  We will definitely be considering this kind of approach for other parts of the UI as well.

**Reply by smandlavdiya** [reply]
"We will definitely be considering this kind of approach for other parts of the UI as well." is it something we can expect soon ? at least adding new pages just like account console

**Reply by jonkoops** [reply]
Yes. https://github.com/keycloak/keycloak/discussions/24805

**Reply by oculos** [reply]
@jonkoops and @ssilvert Thanks for the great work.
I was just wondering - I need to extend the admin theme to add a column to the users's list. I cloned the repo, found the files to add the column, did the modification. But I can't test it. 

I got an error on the development environment - even a blank one (the server app works fine): 

<img width="690" alt="Screenshot 2024-05-30 at 23 06 14" src="https://github.com/keycloak/keycloak/assets/6791923/b958a425-f22d-44fe-9732-1400cd87e540">


And it seems that these jars are quite tightly connected to the version they were built upon. It seems that building the admin-ui.jar on a plain main branch gives me a 999-snapshot that didn't work on the 24.0.4 I have running.

So I'm basically stuck, and I don't know what's the best procedure here to build it and deploy it.

If you could give me a three-four line with the right steps on how to build a useful jar, where to place it, etc, I'd be very grateful.
